source/material_model/viscoelastic.cc

```diff
@@ -0,0 +1,700 @@
+
+          out.thermal_expansion_coefficients[i] = average_value(volume_fractions, thermal_expansivities, arithmetic);
+
+          // Compressibility at the given positions.
+          // The compressibility is given as
+          // $\frac 1\rho \frac{\partial\rho}{\partial p}$.
+          // (here we use an incompressible medium)
+          out.compressibilities[i] = 0.0;
+          // Pressure derivative of entropy at the given positions.
+          out.entropy_derivative_pressure[i] = 0.0;
+          // Temperature derivative of entropy at the given positions.
+          out.entropy_derivative_temperature[i] = 0.0;
+          // Change in composition due to chemical reactions at the
+          // given positions. The term reaction_terms[i][c] is the
+          // change in compositional field c at point i.
+          for (unsigned int c=0; c<in.composition[i].size(); ++c)
+            out.reaction_terms[i][c] = 0.0;
+
+          // Average viscosity
+          const double average_viscosity = calculate_average_viscosity(composition,
+                                                                       viscosities,
+                                                                       viscosity_averaging);
+          // Average elastic shear modulus
+          const double average_elastic_shear_modulus =
+            calculate_average_elastic_shear_modulus(composition,
+                                                    elastic_shear_moduli,
+                                                    viscosity_averaging);
+
+          // Average viscoelastic (e.g., effective) viscosity (equation 28 in Moresi et al., 2003, J. Comp. Phys.)
+          const double average_viscoelastic_viscosity =
+            calculate_average_viscoelastic_viscosity(average_viscosity,
+                                                     average_elastic_shear_modulus,
+                                                     dte);
+
+          out.viscosities[i] = average_viscoelastic_viscosity;
+
+          // Fill elastic outputs if they exist
+          if (ElasticAdditionalOutputs<dim> *elastic_out = out.template get_additional_output<ElasticAdditionalOutputs<dim> >())
+            {
+              elastic_out->elastic_shear_moduli[i] = average_elastic_shear_modulus;
+            }
+
+        }
+
+      // Viscoelasticity section
+      if (in.current_cell.state() == IteratorState::valid && this->get_timestep_number() > 0 && in.strain_rate.size() > 0)
+        {
+          // Get old (previous time step) velocity gradients
+          std::vector<Point<dim> > quadrature_positions(in.position.size());
+          for (unsigned int i=0; i < in.position.size(); ++i)
+            quadrature_positions[i] = this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]);
+
+          FEValues<dim> fe_values (this->get_mapping(),
+                                   this->get_fe(),
+                                   Quadrature<dim>(quadrature_positions),
```

Comment:
Can you add a comment that the feValues needs a quadrature, so we give it a quadrature and we use a default quadrature here because we only want to evaluate the solution (or its gradients) and do not need to integrate anything